From ad4bf46fd81b62d1f8ca237a504c7198986cb9be Mon Sep 17 00:00:00 2001
From: Corentin Oparowski <corentin.oparowski@vates.tech>
Date: Sat, 19 Sep 2015 08:09:42 +0200
Subject: [PATCH 03/36] Several order 4 & 5 allocations are required when
 setting up the interface. These may fail if the interface is reconfigured
 some time after boot due to external memory fragmentation. These buffers must
 be contiguous (for DMA) and they cannot be pre-allocated at boot because they
 are dependent on the number of queues, which is configurable (apart from
 wasting memory if the interface is not up).  Use the __GFP_RETRY_MAYFAIL flag
 when performing these allocations so that direct reclaim and compaction has a
 better chance of working, making the allocations less likely to fail.

Signed-off-by: Ross Lagerwall <ross.lagerwall@citrix.com>
---
 drivers/net/ethernet/broadcom/tg3.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/broadcom/tg3.c b/drivers/net/ethernet/broadcom/tg3.c
index 0acdfd523704f..26dd1c7491985 100644
--- a/drivers/net/ethernet/broadcom/tg3.c
+++ b/drivers/net/ethernet/broadcom/tg3.c
@@ -8599,7 +8599,7 @@ static int tg3_rx_prodring_init(struct tg3 *tp,
 	tpr->rx_std = dma_alloc_coherent(&tp->pdev->dev,
 					 TG3_RX_STD_RING_BYTES(tp),
 					 &tpr->rx_std_mapping,
-					 GFP_KERNEL);
+					 GFP_KERNEL | __GFP_RETRY_MAYFAIL);
 	if (!tpr->rx_std)
 		goto err_out;
 
@@ -8612,7 +8612,7 @@ static int tg3_rx_prodring_init(struct tg3 *tp,
 		tpr->rx_jmb = dma_alloc_coherent(&tp->pdev->dev,
 						 TG3_RX_JMB_RING_BYTES(tp),
 						 &tpr->rx_jmb_mapping,
-						 GFP_KERNEL);
+						 GFP_KERNEL | __GFP_RETRY_MAYFAIL);
 		if (!tpr->rx_jmb)
 			goto err_out;
 	}
@@ -8799,7 +8799,7 @@ static int tg3_mem_rx_acquire(struct tg3 *tp)
 		tnapi->rx_rcb = dma_alloc_coherent(&tp->pdev->dev,
 						   TG3_RX_RCB_RING_BYTES(tp),
 						   &tnapi->rx_rcb_mapping,
-						   GFP_KERNEL);
+						   GFP_KERNEL | __GFP_RETRY_MAYFAIL);
 		if (!tnapi->rx_rcb)
 			goto err_out;
 	}
-- 
2.47.3

